drivers/dma/dma_iproc_pax_v2.c

```diff
@@ -0,0 +1,1058 @@
+					r * PAX_DMA_PER_RING_ALLOC_SIZE);
+		if (!pd->ring[r].ring_mem) {
+			LOG_ERR("RING%d failed to alloc desc memory!\n", r);
+			return -ENOMEM;
+		}
+		/* Find 8K aligned address within allocated region */
+		mem_aligned = ((uintptr_t)pd->ring[r].ring_mem +
+			       PAX_DMA_RING_ALIGN - 1) &
+			       ~(PAX_DMA_RING_ALIGN - 1);
+
+		pd->ring[r].cmpl = (void *)mem_aligned;
+		pd->ring[r].bd = (void *)(mem_aligned +
+					  PAX_DMA_RM_CMPL_RING_SIZE);
+		pd->ring[r].payload = (void *)((uintptr_t)pd->ring[r].bd +
+				      PAX_DMA_RM_DESC_RING_SIZE *
+				      PAX_DMA_NUM_BD_BUFFS);
+
+		LOG_DBG("Ring%d,allocated Mem:0x%p Size %d\n",
+			pd->ring[r].idx,
+			pd->ring[r].ring_mem,
+			PAX_DMA_PER_RING_ALLOC_SIZE);
+		LOG_DBG("Ring%d,BD:0x%p, CMPL:0x%p, PL:0x%p\n",
+			pd->ring[r].idx,
+			pd->ring[r].bd,
+			pd->ring[r].cmpl,
+			pd->ring[r].payload);
+
+		/* Prepare ring desc table */
+		prepare_ring(&(pd->ring[r]));
+
+		/* initialize ring */
+		init_ring(pd, r);
+	}
+
+	/* set ring config done */
+	rm_cfg_finish(pd);
+
+#ifndef CONFIG_DMA_IPROC_PAX_POLL_MODE
+	/* Register and enable RM interrupt */
+	IRQ_CONNECT(DT_INST_IRQN(0),
+		    DT_INST_IRQ(0, priority),
+		    rm_isr,
+		    DEVICE_GET(dma_iproc_pax),
+		    0);
+	irq_enable(DT_INST_IRQN(0));
+#else
+	LOG_INF("%s PAX DMA rings in poll mode!\n", PAX_DMA_DEV_NAME(dev));
+#endif
+	LOG_INF("%s RM setup %d rings\n", PAX_DMA_DEV_NAME(dev),
+		pd->used_rings);
+
+	return 0;
+}
+
+static int dma_iproc_pax_gen_bd(struct dma_iproc_pax_rm_ring_data *ring,
+				struct dma_iproc_pax_payload *pl,
+				uint32_t *pkt_count)
+{
+	uint32_t outstanding, remaining_len;
+	uint32_t offset, curr, mega_len;
+	uint64_t axi_addr;
+	uint64_t pci_addr;
+
+	outstanding = pl->xfer_sz;
+	offset = 0;
+	while (outstanding) {
+		curr = MIN(outstanding, PAX_DMA_MAX_SZ_PER_BD);
+		mega_len = curr / PAX_DMA_MEGA_LENGTH_MULTIPLE;
+		remaining_len = curr % PAX_DMA_MEGA_LENGTH_MULTIPLE;
+		pci_addr = pl->pci_addr + offset;
+		axi_addr = pl->axi_addr + offset;
+
+		if (mega_len) {
+			RM_HEADER_V2(__CURR_INC(ring), curr_toggle_val(ring),
+				     curr_pkt_id(ring), PAX_DMA_RM_DESC_BDCOUNT,
+				     pci_addr);
+			PCIE_DESC(__CURR_INC(ring),
+				  curr_toggle_val(ring), pci_addr);
+			SRC_DST_DESC(__CURR_INC(ring), true,
+				     curr_toggle_val(ring), axi_addr, mega_len,
+				     pl->direction);
+			offset = offset + mega_len *
+				 PAX_DMA_MEGA_LENGTH_MULTIPLE;
+			*pkt_count = *pkt_count + 1;
+		}
+
+		if (remaining_len) {
+			pci_addr = pl->pci_addr + offset;
+			axi_addr = pl->axi_addr + offset;
+			RM_HEADER_V2(__CURR_INC(ring), curr_toggle_val(ring),
+				     curr_pkt_id(ring), PAX_DMA_RM_DESC_BDCOUNT,
+				     pci_addr);
+			PCIE_DESC(__CURR_INC(ring),
+				  curr_toggle_val(ring), pci_addr);
+			SRC_DST_DESC(__CURR_INC(ring), false,
+				     curr_toggle_val(ring), axi_addr,
+				     remaining_len,
+				     pl->direction);
+			offset = offset + remaining_len;
+			*pkt_count = *pkt_count + 1;
+		}
```

Comment:
1) it seems like the code are 90% duplicated in mega_len and non_mega_len case. The main difference is "is_mega" and size/offset calculation. Can we refactor this part inito one 1 function?

2) I am curious: each header can have 15 pair of src and dest. Why we need create extra packets here?

Reply:
1.  Yes, Sure. Let me refactor the code here.

2.  I understand, with our experience on Valkyrie extra header packet didn't impact much on the performance.
     Will try to assess performance difference on Viper and enhance the driver in a separate PR.
 
Incase of Non Mega src/dst descriptors,
Header BD's =  SG list length/15.
Non header BD's = SG list length % 15
So its easy to generate BD's here.

In case of Mega SRC/DST descriptors, Above calculation will not work as each sglist can come with any size and need to be split into different Mega SRC/DST descriptors. So descriptor generation is not straight forward and any additional code to handle will add burden on the CPU.

Reply:
Hi Xin Xie,

I have now added support to handle dma_block_config's list or scatter/gather list handling in dma_iproc_pax_v2.c driver.
And also handling 15 pair of PCIE/AXI descriptors in one header BD. Please review the same.